src/troubleshooting/content.md

```diff
@@ -0,0 +1,641 @@
+
+### CrashLoopBackOff
+
+If the container can't start, then Kubernetes shows the CrashLoopBackOff message as a status.
+
+Usually, a container can't start when:
+
+1. There's an error in the application that prevents it from starting
+1. You [misconfigured the container](https://stackoverflow.com/questions/41604499/my-kubernetes-pods-keep-crashing-with-crashloopbackoff-but-i-cant-find-any-lo)
+1. The Liveness probe failed too many times
+
+You should try and retrieve the logs from that container to investigate why it failed.
+
+If you can't see the logs because your container is restarting too quickly, you can use the following command:
+
+```terminal|command=1|title=bash
+kubectl logs <pod-name> --previous
+```
+
+Which prints the error messages from the previous container.
+
+### RunContainerError
+
+The error appears when the container is unable to start.
+
+That's even before the application inside the container starts.
+
+The issue is usually due to misconfiguration such as:
+
+- mounting a not-existent volume such as ConfigMap or Secrets
+- mounting a read-only volume as read-write
+
+You should use `kubectl describe pod <pod-name>` to collect and analyse the error.
+
+### Pods in a _Pending_ state
+
+When you create a Pod, the Pod stays in the _Pending_ state.
+
+_Why?_
+
+Assuming that your scheduler component is running fine, here are the causes:
+
+1. The cluster doesn't have enough resources such as CPU and memory to run the Pod
+1. The current Namespace has a ResourceQuota object and creating the Pod will make the Namespace go over the quota
+1. The Pod is bound to a _Pending_ PersistentVolumeClaim
+
+Your best option is to inspect the _Events_ section in the `kubectl describe` command:
+
+```terminal|command=1|title=bash
+kubectl describe pod <pod name>
+```
+
+For errors that are created as a result of ResourceQuotas, you can inspect the logs of the cluster with:
+
+```terminal|command=1|title=bash
+kubectl get events --sort-by=.metadata.creationTimestamp
+```
+
+### Pods in a not _Ready_ state
+
+If a Pod is _Running_ but not _Ready_ it means that the Readiness probe is failing.
+
+When the Readiness probe is failing, you the Pods isn't attached to the Service, and no traffic is forwarded to that instance.
+
+A failing Readiness probe is an application-specific error, so you should inspect the _Events_ section in `kubectl describe` to identify the error.
+
+## 2. Troubleshooting Services
+
+If you're Pods are _Running_ and _Ready_, but you're still unable to receive a response from your app, you should check if the Service is configured correctly.
+
+Services are designed to route the traffic to Pods based on their labels.
+
+So the first thing that you should check is how many Pods are target by the Service.
+
+You can do so by checking the Endpoints in the Service:
+
+```terminal|command=1|title=bash
+kubectl describe service <service-name> | grep Endpoints
```

Comment:
there is `kubectl get endpoints`, no need to grep

Reply:
I don't know who designed the endpoint API, but it's rather confusing:
We have:

- Endpoint objects
- endpoints as `<IP>:<port>` pairs

I don't want to introduce the Endpoint object now because that means I have to explain where it comes from.